libnetwork/sandbox_store.go

```diff
@@ -28,9 +28,16 @@ type sbState struct {
 	EpPriority map[string]int
 	// external servers have to be persisted so that on restart of a live-restore
 	// enabled daemon we get the external servers for the running containers.
-	// We have two versions of ExtDNS to support upgrade & downgrade of the daemon
-	// between >=1.14 and <1.14 versions.
-	ExtDNS  []string
+	ExtDNS []extDNSEntry
+	// ExtDNS2 was used to handle migration between docker < 1.14 and >= 1.14.
+	// Before version 1.14 we used ExtDNS but with a []string. As it's unlikely
+	// that installations still have state from before 1.14, we're renaming
+	// ExtDNS2 back to ExtDNS. Keeping both ExtDNS and ExtDNS2 to allow migrating
+	// state from older version.
+	//
+	// Deprecated: only present to migrate existing state. This field will be removed in a future version.
+	//
+	// TODO(thaJeztah): remove this field once we can assume no users have state left from old versions.
```

Comment:
Such users are probably using the old `github.com/docker/libnetwork` repo, so I guess we can just safely remove this field without deprecation period.

Reply:
So (I know it's somewhat confusing); this PR is basically _removing_ the old migration code (which migrated from `ExtDNS -> `ExtDNS2`), and adding a (temporary) migration code to migrate it back (`ExtDNS2 -> ExtDNS`), but now with `ExtDNS` being a `[]extDNSEntry` instead of a `[]string`.

For the migration I added code to

- (temporarily) _write_ the config both to `.ExtDNS` and `.ExtDNS2` (so that a downgrade would still be able to pick up the config)
- (temporarily) check if `ExtDNS` exists, and if not,  fall back to `ExtDNS2` to handle upgrades

Or at least, that's what I attempted to do (be sure to check I didn't mess up) 😅 

We can remove the migration code in 1..2 releases, after which we'll only be using `ExtDNS`